src/gflownet/tasks/qm9/qm9.py

```
@@ -139,86 +99,48 @@ def set_default_hps(self, cfg: Config):
-        else:
-            self.sampling_model = self.model
-        self.algo = TrajectoryBalance(self.env, self.ctx, self.rng, self.cfg)
-
+    def setup_task(self):
         self.task = QM9GapTask(
             dataset=self.training_data,
-            temperature_distribution=self.cfg.task.qm9.temperature_sample_dist,
-            temperature_parameters=self.cfg.task.qm9.temperature_dist_params,
-            num_thermometer_dim=self.cfg.task.qm9.num_thermometer_dim,
+            cfg=self.cfg,
+            rng=self.rng,
             wrap_model=self._wrap_for_mp,
         )
-        self.mb_size = self.cfg.algo.global_batch_size
-        self.clip_grad_param = self.cfg.opt.clip_grad_param
-        self.clip_grad_callback = {
-            "value": lambda params: torch.nn.utils.clip_grad_value_(params, self.clip_grad_param),
-            "norm": lambda params: torch.nn.utils.clip_grad_norm_(params, self.clip_grad_param),
-            "none": lambda x: None,
-        }[self.cfg.opt.clip_grad_type]
-
-    def step(self, loss: Tensor):
-        loss.backward()
-        for i in self.model.parameters():
-            self.clip_grad_callback(i)
-        self.opt.step()
-        self.opt.zero_grad()
-        self.opt_Z.step()
-        self.opt_Z.zero_grad()
-        self.lr_sched.step()
-        self.lr_sched_Z.step()
-        if self.sampling_tau > 0:
-            for a, b in zip(self.model.parameters(), self.sampling_model.parameters()):
-                b.data.mul_(self.sampling_tau).add_(a.data * (1 - self.sampling_tau))
 
 
 def main():
-    """Example of how this model can be run outside of Determined"""
+    """Example of how this model can be run."""
     yaml = YAML(typ="safe", pure=True)
     config_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), "qm9.yaml")
     with open(config_file, "r") as f:
         hps = yaml.load(f)
-    trial = QM9GapTrainer(hps, torch.device("cpu"))
+    trial = QM9GapTrainer(hps, torch.device("cuda"))
```

would make sense to check cuda_availability first.

---

Is it standard to check for cuda and fallback to cpu? I feel like if I was expecting my code to run on cuda and it silently didn't, I'd rather the experiment fail with an exception than silently take 50x longer.

---

It's for making it configurable so someone without access to a GPU would still be able to run it (this could make sense too in case there are integration test to run on github). 

But you are right that failing explicitly is better.

---

Made device configurable 👍